rest/model/conversions.go

```diff
@@ -1,84 +1,95 @@
 package model
 
 import (
+	"fmt"
 	"go/types"
+	"regexp"
+	"strings"
 
 	"github.com/pkg/errors"
 )
 
-type convertType string
-
-func stringToString(in string) string {
-	return in
-}
+const (
+	conversionTemplatePath = "templates/conversions.gotmpl"
+)
 
-func stringToStringPtr(in string) *string {
-	return &in
+type typeInfo struct {
+	Type string
 }
 
-func stringPtrToString(in *string) string {
-	if in == nil {
-		return ""
+// conversionFn uses the specific naming conventions of type conversion funcs in conversions.gotmpl
+// to find the correct generated type conversion function
+func conversionFn(in types.Type, outIsPtr bool, generatedConversions map[string]string) (string, error) {
+	if _, inIsPrimitive := in.(*types.Basic); inIsPrimitive {
+		return noTransforms(in.String(), outIsPtr, generatedConversions)
+	}
+	if intype, inIsPtr := in.(*types.Pointer); inIsPtr {
+		value, isPrimitive := intype.Elem().(*types.Basic)
+		if isPrimitive {
+			return noTransforms(value.String(), outIsPtr, generatedConversions)
+		}
+		return conversionFn(value, outIsPtr, generatedConversions)
+	}
+	if intype, inIsNamedType := in.(*types.Named); inIsNamedType {
+		switch intype.String() {
+		case "time.Time":
+			return noTransforms(intype.String(), outIsPtr, generatedConversions)
+		default:
+			return "", errors.New("complex objects not implemented yet")
+		}
+	}
+	if intype, inIsMap := in.(*types.Map); inIsMap {
+		switch intype.String() {
+		case "map[string]interface{}":
+			return noTransforms(intype.String(), outIsPtr, generatedConversions)
+		default:
+			return "", errors.New("complex maps not implemented yet")
+		}
+	}
+	if intype, inIsInterface := in.(*types.Interface); inIsInterface {
+		switch intype.String() {
+		case "interface{}":
+			return noTransforms(intype.String(), outIsPtr, generatedConversions)
+		default:
+			return "", errors.New("non-empty interfaces not implemented yet")
+		}
 	}
-	return *in
-}
-
-func stringPtrToStringPtr(in *string) *string {
-	return in
-}
-
-func intToInt(in int) int {
-	return in
-}
 
-func intToIntPtr(in int) *int {
-	return &in
+	return "", errors.Errorf("converting type %s is not supported", in.String())
 }
 
-func intPtrToInt(in *int) int {
-	if in == nil {
-		return 0
+func noTransforms(typeName string, outIsPtr bool, generatedConversions map[string]string) (string, error) {
+	if err := generateConversionCode(generatedConversions, typeName); err != nil {
+		return "", err
 	}
-	return *in
+	typeName = cleanName(typeName)
+	if outIsPtr {
+		return fmt.Sprintf("%s%sPtr", typeName, typeName), nil
```

Comment:
Will the in type and out type here always match? In which case do we need to repeat it?

Reply:
This function is only called in scenarios where the in and out types match